src/libtest/lib.rs

```diff
@@ -1087,6 +1099,28 @@ pub fn run_test(opts: &TestOpts,
             let stdout = data.lock().unwrap().to_vec();
             monitor_ch.send((desc.clone(), test_result, stdout)).unwrap();
         });
+
+        // If the thread failed to spawn (for example if the platform doesn't support threads),
+        // we do everything in a single-threaded way instead.
+        if let Err(_) = res {
+            let testfn = testfn2.lock().unwrap().take().unwrap();
+            let mut testfn = panic::AssertRecoverSafe::new(Some(testfn));
+
+            let test_result = if !nocapture {
+                let old_print = io::set_print(box Sink(data3.clone()));
+                let old_panic = io::set_panic(box Sink(data3.clone()));
+                let r = panic::recover(move || (testfn.take().unwrap())());
+                if let Some(old_print) = old_print { io::set_print(old_print); }
+                if let Some(old_panic) = old_panic { io::set_panic(old_panic); }
+                r
+            } else {
+                panic::recover(move || (testfn.take().unwrap())())
+            };
+
+            let test_result = calc_result(&desc2, test_result);
+            let stdout = data3.lock().unwrap().to_vec();
+            monitor_ch2.send((desc2.clone(), test_result, stdout)).unwrap();
+        }
```

Comment:
I don't think it's possible to reduce it, except for some minor things.
It would probably be feasible if `Builder::spawn()` returned the closure that was supposed to be called in the case of an error.